[SYCL][libdevice] Add IMF support in devicelib

jinge90 opened the pull request:
Signed-off-by: jinge90 <ge.jin@intel.com>

To provide support for math APIs and functions, not included in standard <cmath> or <CL/sycl.hpp>.
These math APIs are implemented in "imf" (Intel math functions) SYCL device library. All function names are decorated with \_\_imf_ prefix. These functions may be further categorized into:

1. Different accuracy implementation for some math functions such as log, exp, cos, sin…
2. Type cast util functions for float, double, half, bfloat16
3. Integer type util functions 
4. SIMD util functions
5. Half type util functions
6. Bfloat16 type util functions

This PR is 1st of a series of patches to add “imf” SYCL device library, it includes type cast util functions for float, double and integer type util functions.

akolesov-intel reviewed (commented):
Thank you, JIn Ge, I will look at the IMF part. Try to add me to reviewer using this username.

akolesov-intel commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,428 @@
+#endif
+}
+
+static inline float __fma(float x, float y, float z) {
```
Are you going to add other basic builtins like: __sqrt, __rint, __floor, __ceil, __trunc, __fabs, __fmin, __fmax, probably __rsqrt? All of them are used in IMF math implementations.

jinge90 replied:
Hi, @akolesov-intel 
I added __sqrt, __rint, __floor, __ceil, __trunc, __fabs, __fmin, __fmax for float, double, half in device_imf.hpp and did some simple test in local. For __rsqrt, I think supporting in the same way should be OK, for device code, we have __spirv_ocl_rsqrt but for host code, there is no __builtin_rsqrt, can we use "1.0/sqrt" instead?

Thanks very much.

akolesov-intel replied:
> no __builtin_rsqrt, can we use "1.0/sqrt" instead?

Sure, I do the same for internal testing. Thanks

akolesov-intel commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,428 @@
+#endif
+}
+
+static inline float __fma(float x, float y, float z) {
```
Are you going to add other basic builtins like: __sqrt, __rint, __floor, __ceil, __trunc, __fabs, __fmin, __fmax, probably __rsqrt? All of them are used in IMF math implementations.

akolesov-intel commented on libdevice/imf_utils/saturatef.cpp:
```diff
@@ -0,0 +1,16 @@
+#ifdef __LIBDEVICE_IMF_ENABLED__
+
+DEVICE_EXTERN_C_INLINE
+float __devicelib_imf_saturatef(float x) { return __fclamp(x, .0f, 1.f); }
```
Numerics team owns a number of trivial math functions like __devicelib_imf_copysign, __devicelib_imf_fabs, __devicelib_imf_rsqrt which can be either mapped to just one device independent builtin or implemented by a couple of simple operations - these functions are better to be inlined from header file than called from binary. We would like to gather them all under the ./imf folder which our team owns, promotes and supports. All trivial functions may combined within  just one file. Could you please add such header inside ./imf, say, ./imf/imf_inline.hpp, with one function example (floor?)? Thanks!

jinge90 replied:
Hi, @akolesov-intel 
If we want to inline those trivial functions in header, then the header should be exposed to users, users will include it in their  source code. And in the header file, we need to include:
1. declarations for all imf functions provided by us.
2. Some trivial functions inline definition.

akolesov-intel replied:
I am fine if these functions are going to be inlined as IR, not as sources in header. I am just suggesting to gather them all into one file, not to spread by many separate tiny sources. I did that in my package through imf_inline.cpp source (can be named differently).

jinge90 replied:
Hi, @akolesov-intel 
I updated the patch to integrate all "trivial inline" functions, those functions are inlined as IR.
If we inline in LLVM IR level, we need also split imf_inline.cpp into imf_inline_fp32.cpp and imf_inline_fp64.cpp, I did this in the latest patch and basic tests for fp32, fp64, fp16 passed. Could you help review?
Thanks very much.

akolesov-intel replied:
Thank you, @jinge90 , the split looks good to me.

akolesov-intel commented on libdevice/imf_utils/saturatef.cpp:
```diff
@@ -0,0 +1,16 @@
+#ifdef __LIBDEVICE_IMF_ENABLED__
+
+DEVICE_EXTERN_C_INLINE
+float __devicelib_imf_saturatef(float x) { return __fclamp(x, .0f, 1.f); }
```
Numerics team owns a number of trivial math functions like __devicelib_imf_copysign, __devicelib_imf_fabs, __devicelib_imf_rsqrt which can be either mapped to just one device independent builtin or implemented by a couple of simple operations - these functions are better to be inlined from header file than called from binary. We would like to gather them all under the ./imf folder which our team owns, promotes and supports. All trivial functions may combined within  just one file. Could you please add such header inside ./imf, say, ./imf/imf_inline.hpp, with one function example (floor?)? Thanks!

akolesov-intel commented on libdevice/imf/exp_h_la.cpp:
```diff
@@ -0,0 +1,125 @@
+  return r;
+}
+
+DEVICE_EXTERN_C_INLINE
```
While I believe that suggested approach where each fp16 arithmetic operation is implemented through fp16->fp32->op32->fp16 conversion should work but most possibly it will be presumably slower than if we just convert 16->32, call 32-bit math function as a whole and convert 32 result back to 16: fp16->fp32->exp32->fp16.
So, if we do care about host performance in fp16 case than the second approach looks more preferable on hosts with no fp16 hardware support. Host with fp16 hardware support should map all fp16 arithmetic operations to native 16-bit floating point instructions.

jinge90 replied:
Hi, @akolesov-intel 
I understand fp16->fp32->exp32->fp16 should have better perf than fp16->fp32->op32->fp16. The intention to do "fp16->fp32->op32->fp16" is to share the same implementation  between host device and other device. 
For SYCL host device, performance is not so important, we only need to guarantee functionality since the most important usage for SYCL host device is to provide a reference for other devices(CPU, GPU, FPGA).
Thanks very much.

akolesov-intel replied:
That's fine if performance is not important for host

akolesov-intel commented on libdevice/imf/exp_h_la.cpp:
```diff
@@ -0,0 +1,125 @@
+  return r;
+}
+
+DEVICE_EXTERN_C_INLINE
```
While I believe that suggested approach where each fp16 arithmetic operation is implemented through fp16->fp32->op32->fp16 conversion should work but most possibly it will be presumably slower than if we just convert 16->32, call 32-bit math function as a whole and convert 32 result back to 16: fp16->fp32->exp32->fp16.
So, if we do care about host performance in fp16 case than the second approach looks more preferable on hosts with no fp16 hardware support. Host with fp16 hardware support should map all fp16 arithmetic operations to native 16-bit floating point instructions.

akolesov-intel commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,428 @@
+    *this = (operator float() * static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator/=(const _iml_half &rh) {
```
Have you considered support of hosts with native float16 hardware instructions (SapphireRapids)? Maybe it should be one more branch here.

akolesov-intel commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,428 @@
+    *this = (operator float() * static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator/=(const _iml_half &rh) {
```
Have you considered support of hosts with native float16 hardware instructions (SapphireRapids)? Maybe it should be one more branch here.

jinge90 replied:
Hi, @akolesov-intel 
Yes, we will add it in the future and we still need to keep the half emul in source to support platforms without native FP16.
It seems that we need some dynamic dispatch mechanism like "cpu_dispatch" in source, dispatch to native FP16 implementation only when underlying hardware supports it.

Thanks very much.

akolesov-intel commented on libdevice/imf_wrapper.cpp:
```diff
@@ -0,0 +1,162 @@
+//==----- imf_wrapper.cpp - wrappers for intel math library functions ------==//
```
Maybe it would be more clear to name this file as imf_wrapper_fp32.cpp. And move all float16 related stuff to another file imf_wrapper_fp16.cpp.

jinge90 replied:
Hi, @akolesov-intel 
I agree that splitting all wrapper functions into fp32 and fp16 should be clear in source code and there is no difference in functionality. However, as we can see, the cmake for imf build is very complicated now, the reason is original cmake has some pre-assumption that all functions in the same wrapper/fallback devicelib are in one single source. I have some vague idea to simplify the cmake for imf devicelib and support source split in a clean way but haven't tried. Can we keep fp32 and fp16 in one source currently and split the files when the cmake refactor is done?

akolesov-intel commented on libdevice/imf_wrapper.cpp:
```diff
@@ -0,0 +1,162 @@
+//==----- imf_wrapper.cpp - wrappers for intel math library functions ------==//
```
Maybe it would be more clear to name this file as imf_wrapper_fp32.cpp. And move all float16 related stuff to another file imf_wrapper_fp16.cpp.

akolesov-intel commented on libdevice/imf_wrapper.cpp:
```diff
@@ -0,0 +1,162 @@
+float __imf_saturatef(float x) { return __devicelib_imf_saturatef(x); }
+
+DEVICE_EXTERN_C_INLINE
+float __imf_expf(float x) { return __devicelib_imf_expf(x); }
```
I suppose that this list is going to be expanded as new math functions added under that ./imf folder. Does it make sense to also move it under the ./imf folder in a separate file and give this file ownership to Numerics team? Otherwise we would need to contact Compiler team after every new function addition in order to expand this list.

akolesov-intel replied:
I suppose that this list is going to be expanded as new math functions added under that ./imf folder. Does it make sense to also move it under the ./imf folder in a separate file and give this file ownership to Numerics team? Otherwise we would need to contact Compiler team after every new function addition in order to expand this list.

akolesov-intel commented on libdevice/cmake/modules/SYCLLibdevice.cmake:
```diff
@@ -98,10 +107,168 @@ add_devicelib_obj(libsycl-complex SRC complex_wrapper.cpp DEP ${complex_obj_deps
+set(fallback-imf-src imf_utils/saturatef.cpp
+                     imf_utils/float_convert.cpp
+                     imf_utils/half_convert.cpp
+                     imf/exp_s_la.cpp
```
Just curious is it possible to take the section with all Numerics function out of this file in a separate small cmake include file and move it under the ../imf folder? This case Numerics team is going to be responsible to expand this list after all new function additions and no needs for Compiler team assistance required this case.

akolesov-intel commented on libdevice/cmake/modules/SYCLLibdevice.cmake:
```diff
@@ -98,10 +107,168 @@ add_devicelib_obj(libsycl-complex SRC complex_wrapper.cpp DEP ${complex_obj_deps
+set(fallback-imf-src imf_utils/saturatef.cpp
+                     imf_utils/float_convert.cpp
+                     imf_utils/half_convert.cpp
+                     imf/exp_s_la.cpp
```
Just curious is it possible to take the section with all Numerics function out of this file in a separate small cname include file and move it under the ../imf folder? This case Numerics team is going to be responsible to expand this list after all new function additions and no needs for Compiler team assistance required this case.

jinge90 replied:
Hi, @akolesov-intel 
Moving the section with all Numerics function out of this file in a separate small cmake is OK. However, moving it to ../imf folder doesn't align with libdevice's building system.

xtian-github reviewed (dismissed):
LGTM

jinge90 commented:
Hi, all.
I am focusing on implement type cast functions for float, double in this PR. When those functions are done, I will focus on addressing your comments. Type cast for fp16 and bfloat16 will be in future PR.
Thanks very much for the review.

zettai-reido commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,365 @@
+
+static inline _iml_half_internal __float2half(float x) {
+#if defined(__LIBDEVICE_HOST_IMPL__)
+  uint32_t fp32_bits = *reinterpret_cast<uint32_t *>(&x);
```
If bitcast is available, we should use it here.

zettai-reido commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,365 @@
+  fp32_bits |= sign;
+  fp32_bits |= (exp32 << 23);
+  fp32_bits |= frac32;
+  return *reinterpret_cast<float *>(&fp32_bits);
```
bit_cast?

zettai-reido commented on libdevice/imf/_imf_include.hpp:
```diff
@@ -0,0 +1,86 @@
+  _iml_spbits_t bits;
+  float fp;
+} _iml_sp_union_t;
+inline float as_float(uint32_t a) { return *reinterpret_cast<float *>(&a); }
```
these functions should be replaced by corresponding `__builtin_bit_cast`

jinge90 replied:
Hi, @zettai-reido  and @akolesov-intel 
The __builtin_bit_cast does the same thing as original "reinterpret_cast" code in IR level, if you agree to replace all reinterpret_cast with __builtin_bit_cast for all "as_XXX" functions, I will update the patch.

Thanks very much.

akolesov-intel replied:
We did it internally and observed no problems after replacement. Let's do that in LLVM also please. Thanks!

zettai-reido replied:
Hi @jinge90,
we agreed to replace them.
@akolesov-intel  also tested the latest IMF with them and no issues found.
Thank you!

jinge90 replied:
All as_XX functions are using __builtin_bit_cast now.

zettai-reido commented on libdevice/imf/_imf_include.hpp:
```diff
@@ -0,0 +1,86 @@
+
+inline uint32_t as_uint(float a) { return *reinterpret_cast<uint32_t *>(&a); }
+
+inline double as_double(uint64_t a) { return *reinterpret_cast<double *>(&a); }
```
`__builtin_bit_cast` or analogue

jinge90 replied:
Fixed.

zettai-reido commented on libdevice/imf/_imf_include.hpp:
```diff
@@ -0,0 +1,86 @@
+
+inline double as_double(uint64_t a) { return *reinterpret_cast<double *>(&a); }
+
+inline uint64_t as_ulong(double a) { return *reinterpret_cast<uint64_t *>(&a); }
```
`__builtin_bit_cast` or analogue

jinge90 replied:
Fixed.

zettai-reido commented on libdevice/imf/_imf_include.hpp:
```diff
@@ -0,0 +1,86 @@
+inline uint64_t as_ulong(double a) { return *reinterpret_cast<uint64_t *>(&a); }
+
+static inline _iml_half as_half(uint16_t a) {
+  return *reinterpret_cast<_iml_half *>(&a);
```
`__builtin_bit_cast` or analogue

jinge90 replied:
Fixed.

zettai-reido commented on libdevice/imf/_imf_include.hpp:
```diff
@@ -0,0 +1,86 @@
+}
+
+static inline uint16_t as_short(_iml_half a) {
+  return *reinterpret_cast<uint16_t *>(&a);
```
`__builtin_bit_cast` or analogue

jinge90 replied:
Fixed.

jinge90 commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,428 @@
+#endif
+}
+
+class _iml_half {
```
Hi, @akolesov-intel 
I split the original device_imf.hpp into 2 separate files. All _iml_half related parts are in imf_half.hpp, everything else is still in device_imf.hpp. And device_imf.hpp will include imf_half.hpp, so we only need to include device_imf.hpp in source code.

akolesov-intel replied:
Thank you, @jinge90 . Now it is more clear and easy to find - having all fp16 emulation stuff in the separate file.

jinge90 commented on libdevice/imf/_imf_include.hpp:
```diff
@@ -0,0 +1,86 @@
+  _iml_uint32_t w;
+  float f;
+} int_float;
+typedef union {
```
Hi, @akolesov-intel 
Only typedef won't lead to such problems, keeping one header or split into 2 headers for fp32, fp64 are both OK.

akolesov-intel replied:
I recall that even typedefs with 64-bit data types caused errors on GPU hardware with no 64-bit support, we observed it on MKL VM side. @zettai-reido, can you confirm that?

zettai-reido replied:
Yes, it was a discovery that some `FP64` LLVM IR structures will cause IGC to have (`-11` - build failure)
even in form of compiled typedefs.

jinge90 replied:
Hi, @akolesov-intel and @zettai-reido 
I have split _imf_include.hpp into _imf_include_fp32.hpp and _imf_include_fp64.hpp, could you review?

Thanks very much.

akolesov-intel replied:
Thanks @jinge90 , the change looks good. I will update all math sources accordingly to include 64-bit header only in double precision implementations.

xtian-github commented:
With all the latest changes, the patch LGTM. But, we need to define the promotion/merge process between CE and MKL teams. Thanks.

jinge90 commented on libdevice/imf/_imf_include.hpp:
```diff
@@ -0,0 +1,86 @@
+} _iml_sp_union_t;
+inline float as_float(uint32_t a) { return *reinterpret_cast<float *>(&a); }
+
+inline uint32_t as_uint(float a) { return *reinterpret_cast<uint32_t *>(&a); }
```
Fixed.

xtian-github commented:
@jinge90 *fp32.hpp and *fp64.hpp splitting is good.  as fp64 does not work for ATS-M.

zettai-reido reviewed (dismissed):
Thanks for addressing the comments.
LGTM as it of now.

mdtoguchi commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4714,10 +4714,9 @@ class OffloadingActionBuilder final {
-                                                   {"internal", true}};
+      llvm::StringMap<bool> devicelib_link_info = {
+          {"libc", true},        {"libm-fp32", true},   {"libm-fp64", true},
+          {"libimf-fp32", true}, {"libimf-fp64", true}, {"internal", true}};
```
Could you add tests verifying the addition of the new devicelibs during compilation?

jinge90 replied:
Hi, @mdtoguchi 
The test for imf devicelib have been already added. Could you review again?

Thanks very much.

xtian-github reviewed (approved):
LGTM

xtian-github reviewed (dismissed):
LGTM

jinge90 commented on clang/lib/Driver/ToolChains/MSVC.cpp:
```diff
@@ -139,6 +139,7 @@ void visualstudio::Linker::ConstructJob(Compilation &C, const JobAction &JA,
       CmdArgs.push_back("-defaultlib:sycld.lib");
     else
       CmdArgs.push_back("-defaultlib:sycl.lib");
+    CmdArgs.push_back("-defaultlib:sycl-devicelib-host.lib");
```
Hi, @mdtoguchi 
The library "sycl-devicelib-host.lib" contains implementation for imf device library functions for SYCL host device. This static library will only be used when users are using sycl compiler.

Thanks very much.

jinge90 commented on clang/lib/Driver/ToolChains/Gnu.cpp:
```diff
@@ -714,6 +714,7 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
       if (Args.hasArg(options::OPT_fsycl) &&
           !Args.hasArg(options::OPT_nolibsycl)) {
         CmdArgs.push_back("-lsycl");
+        CmdArgs.push_back("-lsycl-devicelib-host");
```
Hi, @mdtoguchi 
libsycl-devicelib-host.a contains implementation for imf device library for SYCL host device, it is linked when user is using sycl compiler.

Thanks very much.

mdtoguchi reviewed (dismissed):
LGTM, Thanks

asudarsa commented:
Hi @jinge90 
I will report my feedback before EOD today. 
Thanks

akolesov-intel reviewed (approved):
Looks good, thank you!

asudarsa reviewed (dismissed):
sycl-post-link changes look good to me. Thanks.

jinge90 commented:
Hi, @intel/llvm-reviewers-runtime 
Could you review the runtime part of this patch, we follow existing mechanism to add 2 new fallback device libraries: imf-fp32 and imf-fp64.

Thanks very much.

steffenlarsen commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,527 @@
+static inline constexpr To __bit_cast(const From &from) {
+  static_assert(sizeof(To) == sizeof(From),
+                "Can't do bit cast between 2 types with different sizes!");
+  /*static_assert(std::is_trivially_copyable<From>::value &&
+                    std::is_trivially_copyable<To>::value,
+                "Can't do bit cast for type which is not trivially
+     copyable!");*/
```
Should this be replaced with a TODO?

jinge90 replied:
Done.

steffenlarsen commented on libdevice/device_imf.hpp:
```diff
@@ -0,0 +1,527 @@
+static inline float __fclamp(float x, float y, float z) {
+#if defined(__LIBDEVICE_HOST_IMPL__)
+  return __builtin_fmin(__builtin_fmax(x, y), z);
+#elif defined(__SPIR__)
```
This and a bunch of the other definitions will fail if neither `__LIBDEVICE_HOST_IMPL__` nor `__SPIR__` is defined. Could this have a `static_assert` at the beginning of the file to warn in the case that this happens?

jinge90 replied:
Hi, @steffenlarsen 
Code in this file is guarded by another macro "__LIBDEVICE_IMF_ENABLED": https://github.com/intel/llvm/pull/5999/files#diff-9cbc9200719f1499b106121cbcb9010e4a8b5629983bd18e185696a8aea49d67R17
And this marco is defined in device.h when either __SPIR__ or __LIBDEVICE_HOST_IMPL__ is defined.
So, we can add a check in device_imf.hpp when the macro "__LIBDEVICE_IMF_ENABLED" is not defined, is that OK?

Thanks very much.

steffenlarsen replied:
My primary worry is if somehow we unintentionally break this guarantee in the future. We should hopefully get warnings, but I am hoping a `static_assert` at the top would help whoever breaks it understand the reason for the failures faster than having to dig through a pile of missing return errors.

steffenlarsen replied:
This has been addressed.

jinge90 commented on libdevice/device_imf.hpp:
```diff
@@ -16,6 +16,11 @@
 
 #ifdef __LIBDEVICE_IMF_ENABLED__
 
+#if !defined(__SPIR__) && !defined(__LIBDEVICE_HOST_IMPL__)
```
Hi, @steffenlarsen 
Does this change meet your expectation?
Thanks very much.

steffenlarsen replied:
Yes, looks good. Thank you!

steffenlarsen reviewed (approved):
LGTM!

jinge90 commented:
Hi, all.
We made a little change in comments and added one compile-time macro check, could you help review again? The patch is ready to merge now.

Thanks very much.

pvchupin commented:
@jinge90, can you please add PR description
@asudarsa, feel free to reapprove.

asudarsa reviewed (dismissed):
LGTM. Thanks

xtian-github commented:
@jinge90   is @zettai-reido going to do merge? or you?

pvchupin commented:
@xtian-github, I'm going to merge as soon as somebody update PR description above (to be used as commit message).

xtian-github commented:
@pvchupin   Thanks Pavel!

jinge90 commented:
Hi, @pvchupin 
I just updated the PR description and also made a little change to the patch, since the latest change only includes some little bugfix, it is OK to merge.

Thanks very much.

zettai-reido reviewed (approved):
LGTM
